### PR TITLE
[Paddle] Support recompute

### PR DESCRIPTION
**Pull request by Tom-Zheng**
This PR supports recompute for TE Paddle. The user can use `te.recompute` to wrap TE layers, or use `recompute_core_attention=True` in `TransformerLayer.forward` to recompute only the core attention part.

### CODE REVIEW

**Comment by Tom-Zheng** [comment]
+ @jeng1220 to take a look.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by Tom-Zheng** [comment]
@timmoon10 There were some changes since your last approval, would you please take a look?  cc: @jeng1220

**Comment by Tom-Zheng** [comment]
**Changes since last approval**:
1. Improve recompute test: added check of memory usage
2. Improve recompute detection performance: `inspect.stack()` is slow (and messes with tensor lifetime). Use `frame.f_back` instead.
3. Add documentation for `rng_name` args introduced in #410 
4. Other functional fixes

File: tests/paddle/test_recompute.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        optimizer.clear_grad()
+
+    print("Loss: ", float(loss))
+    print("Peak memory: ", paddle.device.cuda.max_memory_allocated(0))
+
+if __name__ == "__main__":
+    main()
+"""
+            # Create 'script.py' file
+            with open('script.py', 'w', encoding="utf8") as script_file:
```
Comment (jeng1220) [inline tests/paddle/test_recompute.py:87]: Just create a file for the code instead of using a string.

File: tests/paddle/test_recompute.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                script_file.write(script_content)
```
Comment (jeng1220) [inline tests/paddle/test_recompute.py:88]: Just create a file for the code instead of using a string.

Comment (Tom-Zheng) [reply tests/paddle/test_recompute.py]: Done

**Comment by jeng1220** [comment]
LGTM.

**Comment by jeng1220** [comment]
@timmoon10 ,

There are some UT failures but they are not related to this PR:
- te-ci/L0_jax_unittest--A100_1GPU
- te-ci/L0_jax_unittest--H100_1GPU
- te-ci/L0_jax_unittest--V100_1GPU
- te-ci/L0_unittest--H100_1GPU

Please review and merge the code if everything looks good.
Thank you.

**Review by timmoon10 — APPROVED** [review]

**Pull request merged**